kong/clustering/config_helper.lua

```diff
@@ -0,0 +1,230 @@
+local constants = require("kong.constants")
+local declarative = require("kong.db.declarative")
+
+local tostring = tostring
+local assert = assert
+local type = type
+local error = error
+local pairs = pairs
+local ipairs = ipairs
+local concat = table.concat
+local sort = table.sort
+
+local isempty = require("table.isempty")
+local isarray = require("table.isarray")
+local nkeys = require("table.nkeys")
+local new_tab = require("table.new")
+
+local ngx_log = ngx.log
+local ngx_null = ngx.null
+local ngx_md5 = ngx.md5
+local ngx_md5_bin = ngx.md5_bin
+
+local ngx_DEBUG = ngx.DEBUG
+
+local DECLARATIVE_EMPTY_CONFIG_HASH = constants.DECLARATIVE_EMPTY_CONFIG_HASH
+local _log_prefix = "[clustering] "
+
+
+local _M = {}
+
+
+local function to_sorted_string(value)
+  if value == ngx_null then
+    return "/null/"
+  end
+
+  local t = type(value)
+  if t == "string" or t == "number" then
+    return value
+  end
+
+  if t == "boolean" then
+    return tostring(value)
+  end
+
+  if t == "table" then
+    if isempty(value) then
+      return "{}"
+    end
+
+    if isarray(value) then
+      local count = #value
+      if count == 1 then
+        return to_sorted_string(value[1])
+      end
+
+      if count == 2 then
+        return to_sorted_string(value[1]) .. ";" ..
+               to_sorted_string(value[2])
+      end
+
+      if count == 3 then
+        return to_sorted_string(value[1]) .. ";" ..
+               to_sorted_string(value[2]) .. ";" ..
+               to_sorted_string(value[3])
+      end
+
+      if count == 4 then
+        return to_sorted_string(value[1]) .. ";" ..
+               to_sorted_string(value[2]) .. ";" ..
+               to_sorted_string(value[3]) .. ";" ..
+               to_sorted_string(value[4])
+      end
+
+      if count == 5 then
+        return to_sorted_string(value[1]) .. ";" ..
+               to_sorted_string(value[2]) .. ";" ..
+               to_sorted_string(value[3]) .. ";" ..
+               to_sorted_string(value[4]) .. ";" ..
+               to_sorted_string(value[5])
+      end
+
+      local i = 0
+      local o = new_tab(count < 100 and count or 100, 0)
+      for j = 1, count do
+        i = i + 1
+        o[i] = to_sorted_string(value[j])
+
+        if j % 100 == 0 then
+          i = 1
+          o[i] = ngx_md5_bin(concat(o, ";", 1, 100))
+        end
```

Comment:
are there plans to fix this off-by-one error?

on every `j==n*100`, replaces all values in `o` with a hash of the first 100 and sets `i` to 1, so the next value (number 101) will go on `o[2]`.  So when `j==200`, there will be 101 values in `o` (the last hash plus 100 values more) but it will again use only 100 values for the next hash, so it drops the hashes values of items number 201, 301, 401.... etc.

Reply:
maybe we can use `local i = 1` and `new_tab(count < 100 and count +1 or 101, 0)`

Reply:
> Shall we fix it?

I don't know.  mentioned it just that it can be discussed.  probably in a separate PR.

Reply:
I think perhaps we can not change it because of the compatibility of old version.

Reply:
the hash method has changed several times already.  there's no compatibility issue

Reply:
Let's fix the off-by-one error on this PR then then. Another PR will make everything take even longer.

Reply:
It will force us to modify all the hashes on the test, which is a pain. 

New directive: Do it only if it can be done quickly on this PR. We have lived with the problem so far.

Reply:
I think we can do it later, now we should finish refactoring first.